### ISSUE TITLE
Wrong invariant inference for frozen dataclass attribute

### ISSUE THREAD

**Issue by light-scattering-io** [microsoft/pyright#9241]
# Bug description

In the example below, I would expect the type variable `T` for the `content` attribute to be covariant and thus `Wrapper[tuple[int, ...]]` can be passed to the `add` function.
It indeed passes without error if passed as a direct argument to the function, but it fails when introducing an intermediate variable as `T` is then estimated to be invariant.

```python
from collections.abc import Sequence
from dataclasses import dataclass


@dataclass(frozen=True)
class Wrapper[T]:
    content: T


def add(value: Wrapper[Sequence[int]]):
    return sum(value.content)


add(Wrapper((1, 2, 3)))  # Fine, no error

x = Wrapper((1, 2, 3))
add(x)  # Error

```

This is the error for the second call to `add`:
```
test.py:17:5 - error: Argument of type "Wrapper[tuple[Literal[1], Literal[2], Literal[3]]]" cannot be assigned to parameter "value" of type "Wrapper[Sequence[int]]" in function "add"
    "Wrapper[tuple[Literal[1], Literal[2], Literal[3]]]" is not assignable to "Wrapper[Sequence[int]]"
      Type parameter "T@Wrapper" is invariant, but "tuple[Literal[1], Literal[2], Literal[3]]" is not the same as "Sequence[int]" (reportArgumentType)
1 error, 0 warnings, 0 informations
```

As a workaround, everything seems fine when using the old `TypeVar` syntax and forcing `covariant=True` for `T`.

# Version

On my local computer (windows 11, python 3.12.3), the error is present for pyright 1.1.384 and 1.1.385, but goes away for 1.1.383.
I have trouble reproducing the error in pyright playground, it doesn't trigger when using python 3.12, but occurs for 3.13.

**Comment by erictraut** [comment]
This is a duplicate of #9226.

**Comment by light-scattering-io** [comment]
Thanks for the answer, I missed the previous report.

Just wondering, if there a reason why the error is only reported when introducing an intermediate variable and not when passing the object directly?

**Comment by erictraut** [comment]
> [Is] there a reason why the error is only reported when introducing an intermediate variable and not when passing the object directly?

Yes, this is because of [bidrectional type inference](https://microsoft.github.io/pyright/#/type-inference?id=bidirectional-type-inference-expected-types), which is applied in the first case but not the second.
